### PR TITLE
MetaX C550: vLLM 0.20.2 adaptation patches

### PR DESCRIPTION
**Pull request by physics31415926**
## MetaX C550: vLLM 0.20.2 Adaptation Patches

Qwen3-4B TP=2 offline inference verified on MetaX C550 (8x 64GB).
Tested on two images:
- Old: `maca.ai3.5.3.20-torch2.8-py310` (MACA 2.33.0, torch 2.8.0+metax3.5.3.9)
- New: `maca.ai3.7.0.105-torch2.8-py310` (torch 2.8.0+metax3.7.0.7)

### Architecture

```
metax.yaml (dispatch config)
├── attention_backend: vendor:metax  → MacaBackend (flash_attn)
├── silu_and_mul:     flagos → reference
├── rms_norm:         flagos → reference
├── rotary_embedding: flagos → reference
└── topk_softmax:     flagos → reference
```

- **MetaX vendor backend**: Only provides `attention_backend` (flash_attn routing)
- **FlagGems (flagos)**: Triton-based ops (silu_and_mul, rms_norm, rotary_embedding, topk_softmax)
- **Reference**: PyTorch native fallback when FlagGems unavailable

### Changes

**Vendor backend (MetaX-specific)**:
- `attention_backend`: Routes to flash_attn/flashmla based on model config
- Removed all other vendor ops (now handled by dispatch system)

**Runtime patches** (commits 7150464, 46a8aca, 55e1c01):
- `topk_topp_sampler.py`: Bypass Triton topk_topp kernel (PassManager::run fails on MetaX Triton)
- `accelerator_compat.py`: Patch `torch.accelerator.empty_cache` (missing in torch 2.8+metax)
- `fa_utils.py`: Pure PyTorch `reshape_and_cache_flash` (vllm._C not compiled)

**Stub _C op registration** (commit 69afb4d):
- `_op_schemas.py`: Hardcoded 78 op schemas from vLLM 0.20.2 `torch_bindings.cpp`
- Works with both editable (`-e`) and wheel installs (no dependency on source tree)

**NCCL/MCCL mapping** (TP>1 support):
- `pynccl_wrapper.py`, `cuda_wrapper.py`, `utils_patch.py`: Map NCCL→MCCL for MetaX

### Test Results

| Test | Image | Result | Notes |
|------|-------|--------|-------|
| Unit tests (349) | old | ✅ Pass | |
| Offline inference Qwen3-4B TP=2 (eager) | old | ✅ Pass | 2.80 tok/s with FlagGems |
| Qwen3-4B TP=2 + torch.compile + CUDAGraph | old | ✅ Pass | **84.34 tok/s** (batch=10, after warmup) |
| Qwen3-4B TP=2 + torch.compile + CUDAGraph | new | ✅ Pass | **56.13 tok/s** (batch=10, after warmup) |
| Qwen3.5-35B-A3B TP=8 (eager) | old | ✅ Pass | **6.09 tok/s** (batch=10, MoE+GDN+Mamba hybrid) |

#### Qwen3-4B (Dense), TP=2, torch.compile + CUDAGraph (FULL_AND_PIECEWISE)

10 prompts × 128 tokens, after 3 warmup rounds.

| Image | Warmup (single prompt) | Batch Throughput (10×128) |
|-------|------------------------|--------------------------|
| Old (`ai3.5.3.20`, torch 2.8.0+metax3.5.3.9) | 73.34 tok/s | **84.34 tok/s** |
| New (`ai3.7.0.105`, torch 2.8.0+metax3.7.0.7) | 70.56 tok/s | **56.13 tok/s** |

#### Qwen3.5-35B-A3B (Hybrid MoE: GatedDeltaNet + Mamba + Attention), TP=8, eager mode

10 prompts × 128 tokens, old image (`ai3.5.3.20`).

| Phase | Tokens | Time | Speed |
|-------|--------|------|-------|
| Warmup 1 (cold, Triton JIT) | 20 | 192.91s | 0.10 tok/s |
| Warmup 2 | 20 | 9.78s | 2.04 tok/s |
| Warmup 3 | 20 | 9.50s | 2.11 tok/s |
| **Batch (10×128)** | **1229** | **201.96s** | **6.09 tok/s** |

FlagGems MoE Triton kernels dispatched: `moe_align_block_size`, `invoke_fused_moe_triton_kernel`, `moe_sum`.

**Key findings**:
- Both images work correctly with torch.compile + CUDAGraph + TP=2 (Qwen3-4B)
- MCCL stream capture conflict (previously expected with TP>1) **did not occur** — vLLM's piecewise CUDAGraph correctly handles all_reduce as splitting op
- Old image is ~50% faster in batch throughput (84 vs 56 tok/s); new image may have regression in MACA runtime or Triton codegen
- FlagGems Triton kernels + torch.compile + CUDAGraph provide **300x+** speedup over PyTorch reference (~0.26 tok/s)
- Qwen3.5-35B-A3B (hybrid MoE) works with eager mode on TP=8; torch.compile fails due to MetaX torch._dynamo guard bug with `hasattr()` on `nn.Module`

### Known Limitations

- `reshape_and_cache_flash` uses PyTorch fallback (no Triton impl yet)
- `topk_topp` sampling kernel bypassed (MetaX Triton compilation failure)
- Qwen3.5-35B-A3B requires `enforce_eager=True` (MetaX torch._dynamo guard bug with `hasattr()` on `nn.Module`)

### FlagGems Feedback (missing ops for MetaX)

1. `reshape_and_cache_flash` — KV cache scatter write
2. `topk_topp` — combined top-k + top-p sampling kernel

### Dependencies

- vLLM 0.20.2 (CPU-only wheel, `VLLM_TARGET_DEVICE=empty`, `--no-deps`)
- torch 2.8.0+metax (vendor-provided, not overwritten)
- flash_attn (MetaX build, pre-installed in image)
- FlagGems 5.0.2 (for Triton ops)

### CODE REVIEW

File: vllm_fl/dispatch/backends/vendor/metax/impl/activation.py [hunk 1]
```diff
@@ -37,5 +38,9 @@ def gelu_and_mul_maca(obj, x: torch.Tensor) -> torch.Tensor:
     Returns:
         Output tensor of shape [..., d]
     """
-    act_fn = GeluAndMul()
-    return act_fn.forward_cuda(x)
+    from flag_gems.fused import gelu_and_mul
```
Comment (cyber-pioneer) [inline vllm_fl/dispatch/backends/vendor/metax/impl/activation.py]: vendor dispatch call gems op?

Comment (physics31415926) [reply vllm_fl/dispatch/backends/vendor/metax/impl/activation.py]: use torch native instead

File: vllm_fl/dispatch/backends/vendor/metax/impl/fused_moe.py
Comment (cyber-pioneer) [inline vllm_fl/dispatch/backends/vendor/metax/impl/fused_moe.py:1]: why delete these files?

Comment (physics31415926) [reply vllm_fl/dispatch/backends/vendor/metax/impl/fused_moe.py:1]: independent from maca libs
